File: backend/app/schemas/product.py
from pydantic import BaseModel, ConfigDict, field_validator
from datetime import datetime
from typing import Optional, List
import json

from app.schemas.category import CategoryResponse


# Shared base 
class ProductBase(BaseModel):
    title: str
    description: Optional[str] = None
    brand: Optional[str] = None
    price: float
    original_price: Optional[float] = None
    discount_percent: float = 0.0
    stock: int = 0
    is_active: bool = True
    image_url: Optional[str] = None
    images: List[str] = []
    rating: float = 0.0
    rating_count: int = 0
    specifications: Optional[str] = None   
    category_id: Optional[int] = None


# Write schemas 

class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    """All fields optional for PATCH semantics."""
    title: Optional[str] = None
    description: Optional[str] = None
    brand: Optional[str] = None
    price: Optional[float] = None
    original_price: Optional[float] = None
    discount_percent: Optional[float] = None
    stock: Optional[int] = None
    is_active: Optional[bool] = None
    image_url: Optional[str] = None
    images: Optional[List[str]] = None
    rating: Optional[float] = None
    rating_count: Optional[int] = None
    specifications: Optional[str] = None
    category_id: Optional[int] = None


# Read schemas 
class ProductCardResponse(BaseModel):
    """Lightweight projection used for listing/grid views."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    brand: Optional[str]
    price: float
    original_price: Optional[float]
    discount_percent: float
    image_url: Optional[str]
    rating: float
    rating_count: int
    stock: int
    category: Optional[CategoryResponse] = None


class ProductDetailResponse(ProductBase):
    """Full projection used for the detail page."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    category: Optional[CategoryResponse] = None
    created_at: datetime

    @field_validator("specifications", mode="before")
    @classmethod
    def parse_specs(cls, v):
        """Return raw string as-is; frontend parses JSON."""
        return v


# Paginated list wrapper

class ProductListResponse(BaseModel):
    items: List[ProductCardResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/types/product.ts
//  Domain Types 

export interface Category {
  id: number;
  name: string;
  slug: string;
  description: string | null;
  icon_url: string | null;
  is_active: boolean;
  created_at: string;
}

/** Lightweight shape returned in listing/grid views */
export interface ProductCard {
  id: number;
  title: string;
  brand: string | null;
  price: number;
  original_price: number | null;
  discount_percent: number;
  image_url: string | null;
  rating: number;
  rating_count: number;
  stock: number;
  category: Category | null;
}

/** Full shape returned on the detail page */
export interface ProductDetail extends ProductCard {
  description: string | null;
  images: string[];
  specifications: string | null;   // raw JSON string; parse on use
  is_active: boolean;
  created_at: string;
}

// API Response Types 

export interface ProductListResponse {
  items: ProductCard[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// Query Param Types 

export type SortOption = "created_at" | "price" | "rating" | "discount_percent";
export type SortOrder = "asc" | "desc";

export interface ProductFilters {
  search?: string;
  category_id?: number;
  min_price?: number;
  max_price?: number;
  sort_by?: SortOption;
  order?: SortOrder;
  page?: number;
  page_size?: number;
}